### ISSUE TITLE
vfs: deadlock on accessing a directory with --poll-changes

### ISSUE THREAD

**Issue by ncw** [ncw/rclone#2811]
#### What is the problem you are having with rclone?

Sometimes the VFS deadlocks when
- multiple programs are reading / writing a directory
- and a directory change comes in

See: https://forum.rclone.org/t/vfs-mount-freezes/7566

#### What is your rclone version (output from `rclone version`)

latest master

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

linux 64 bit

####  Which cloud storage system are you using? (eg Google Drive)

google drive

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

To reproduce

Run this in one window

    rclone mount -vv --poll-interval 0.2s --vfs-cache-mode writes --dir-cache-time=720h drive:test /mnt/tmp/

And this in another

    export i=1 ; while true ; do ls /mnt/tmp/test-deadlock ; echo "hello" > /mnt/tmp/test-deadlock/$i ; rm /mnt/tmp/test-deadlock/$i ; done

And this in a 3rd

    export i=2 ; while true ; do ls /mnt/tmp/test-deadlock ; echo "hello" > /mnt/tmp/test-deadlock/$i ; rm /mnt/tmp/test-deadlock/$i ; done

This deadlocks quite soon after the `i=2` command is run.

It doesn't deadlock if only the `i=1` command is run.

It doesn't deadlock if the `--poll-interval` is set very large.

The deadlock looks like this

```
2018/12/02 22:06:53 DEBUG : Google drive root 'test': Checking for changes on remote
2018/12/02 22:06:53 DEBUG : Google drive root 'test': Checking for changes on remote
2018/12/02 22:06:54 DEBUG : Google drive root 'test': Checking for changes on remote
2018/12/02 22:06:54 DEBUG : test-deadlock/1: Removed from cache
2018/12/02 22:06:54 DEBUG : test-deadlock/: >Remove: err=<nil>
2018/12/02 22:06:54 DEBUG : fuse: -> [ID=0xa7] Remove
2018/12/02 22:06:54 DEBUG : fuse: <- Forget [ID=0xa8 Node=0x5 Uid=0 Gid=0 Pid=0] 1
2018/12/02 22:06:54 DEBUG : fuse: -> [ID=0xa8] Forget
2018/12/02 22:06:54 DEBUG : fuse: <- Lookup [ID=0xa9 Node=0x1 Uid=502 Gid=502 Pid=11555] "test-deadlock"
2018/12/02 22:06:54 DEBUG : /: Lookup: name="test-deadlock"
2018/12/02 22:06:54 DEBUG : /: >Lookup: node=test-deadlock/, err=<nil>
2018/12/02 22:06:54 DEBUG : test-deadlock/: Attr: 
2018/12/02 22:06:54 DEBUG : test-deadlock/: >Attr: attr=valid=1s ino=0 size=0 mode=drwxrwxr-x, err=<nil>
2018/12/02 22:06:54 DEBUG : fuse: -> [ID=0xa9] Lookup 0x5 gen=38 valid=1s attr={valid=1s ino=2334695540606880430 size=0 mode=drwxrwxr-x}
2018/12/02 22:06:54 DEBUG : fuse: <- Forget [ID=0xaa Node=0x5 Uid=0 Gid=0 Pid=0] 1
2018/12/02 22:06:54 DEBUG : fuse: <- Lookup [ID=0xab Node=0x1 Uid=502 Gid=502 Pid=11555] "test-deadlock"
2018/12/02 22:06:54 DEBUG : /: Lookup: name="test-deadlock"
2018/12/02 22:06:54 DEBUG : /: >Lookup: node=test-deadlock/, err=<nil>
2018/12/02 22:06:54 DEBUG : test-deadlock/: Attr: 
2018/12/02 22:06:54 DEBUG : test-deadlock/: >Attr: attr=valid=1s ino=0 size=0 mode=drwxrwxr-x, err=<nil>
2018/12/02 22:06:54 DEBUG : fuse: -> [ID=0xab] Lookup 0x5 gen=39 valid=1s attr={valid=1s ino=2334695540606880430 size=0 mode=drwxrwxr-x}
2018/12/02 22:06:54 DEBUG : fuse: <- Forget [ID=0xac Node=0x2 Uid=0 Gid=0 Pid=0] 1
2018/12/02 22:06:54 DEBUG : fuse: -> [ID=0xac] Forget
2018/12/02 22:06:54 DEBUG : fuse: -> [ID=0xaa] Forget
2018/12/02 22:06:54 DEBUG : fuse: <- Open [ID=0xad Node=0x5 Uid=502 Gid=502 Pid=11555] dir=true fl=OpenReadOnly+OpenDirectory+OpenNonblock
2018/12/02 22:06:54 DEBUG : fuse: -> [ID=0xad] Open 0x1 fl=0
2018/12/02 22:06:54 DEBUG : fuse: <- Read [ID=0xae Node=0x5 Uid=502 Gid=502 Pid=11555] 0x1 4096 @0x0 dir=true fl=0 lock=0 ffl=OpenReadOnly+OpenDirectory+OpenNonblock
2018/12/02 22:06:54 DEBUG : test-deadlock/: ReadDirAll: 
2018/12/02 22:06:54 DEBUG : test-deadlock/: >ReadDirAll: item=1, err=<nil>
2018/12/02 22:06:54 DEBUG : fuse: -> [ID=0xae] Read 32
2018/12/02 22:06:54 DEBUG : fuse: <- Lookup [ID=0xaf Node=0x5 Uid=502 Gid=502 Pid=11555] "2"
2018/12/02 22:06:54 DEBUG : test-deadlock/: Lookup: name="2"
2018/12/02 22:06:54 DEBUG : test-deadlock/: >Lookup: node=test-deadlock/2, err=<nil>
2018/12/02 22:06:54 DEBUG : test-deadlock/2: Attr: 
2018/12/02 22:06:54 DEBUG : test-deadlock/2: >Attr: a=valid=1s ino=0 size=6 mode=-rw-rw-r--, err=<nil>
2018/12/02 22:06:54 DEBUG : fuse: -> [ID=0xaf] Lookup 0x2 gen=40 valid=1s attr={valid=1s ino=4004752764195615492 size=6 mode=-rw-rw-r--}
2018/12/02 22:06:54 DEBUG : fuse: <- Read [ID=0xb0 Node=0x5 Uid=502 Gid=502 Pid=11555] 0x1 4096 @0x20 dir=true fl=0 lock=0 ffl=OpenReadOnly+OpenDirectory+OpenNonblock
2018/12/02 22:06:54 DEBUG : fuse: -> [ID=0xb0] Read 0
2018/12/02 22:06:54 DEBUG : fuse: <- Release [ID=0xb1 Node=0x5 Uid=0 Gid=0 Pid=0] 0x1 fl=OpenReadOnly+OpenDirectory+OpenNonblock rfl=0 owner=0x0
2018/12/02 22:06:54 DEBUG : fuse: -> [ID=0xb1] Release
2018/12/02 22:06:54 DEBUG : fuse: <- Lookup [ID=0xb2 Node=0x5 Uid=502 Gid=502 Pid=32351] "1"
2018/12/02 22:06:54 DEBUG : test-deadlock/: Lookup: name="1"
2018/12/02 22:06:54 DEBUG : test-deadlock/: >Lookup: node=<nil>, err=no such file or directory
2018/12/02 22:06:54 DEBUG : fuse: -> [ID=0xb2] Lookup error=ENOENT
2018/12/02 22:06:54 DEBUG : fuse: <- Create [ID=0xb3 Node=0x5 Uid=502 Gid=502 Pid=32351] "1" fl=OpenWriteOnly+OpenCreate+OpenTruncate mode=-rw-rw-r-- umask=--------w-
2018/12/02 22:06:54 DEBUG : test-deadlock/: Create: name="1"
2018/12/02 22:06:54 DEBUG : test-deadlock/1: Open: flags=O_WRONLY|O_CREATE|O_TRUNC
2018/12/02 22:06:54 DEBUG : test-deadlock/1(0xc0000bdb00): Opening cached copy with flags=O_WRONLY|O_CREATE|O_TRUNC
2018/12/02 22:06:54 DEBUG : test-deadlock/1: >Open: fd=test-deadlock/1 (rw), err=<nil>
2018/12/02 22:06:54 DEBUG : test-deadlock/: >Create: node=test-deadlock/1, handle=&{test-deadlock/1 (rw)}, err=<nil>
2018/12/02 22:06:54 DEBUG : test-deadlock/1: Attr: 
2018/12/02 22:06:54 DEBUG : test-deadlock/1: >Attr: a=valid=1s ino=0 size=0 mode=-rw-rw-r--, err=<nil>
2018/12/02 22:06:54 DEBUG : fuse: -> [ID=0xb3] Create {0x3 gen=41 valid=1m0s attr={valid=1s ino=4004756062730500125 size=0 mode=-rw-rw-r--}} {0x1 fl=0}
2018/12/02 22:06:54 DEBUG : fuse: <- Flush [ID=0xb4 Node=0x3 Uid=502 Gid=502 Pid=32351] 0x1 fl=0x0 lk=0xb8ca17e21032f904
2018/12/02 22:06:54 DEBUG : &{test-deadlock/1 (rw)}: Flush: 
2018/12/02 22:06:54 DEBUG : test-deadlock/1(0xc0000bdb00): RWFileHandle.Flush ignoring flush on unwritten handle
2018/12/02 22:06:54 DEBUG : &{test-deadlock/1 (rw)}: >Flush: err=<nil>
2018/12/02 22:06:54 DEBUG : fuse: -> [ID=0xb4] Flush
2018/12/02 22:06:54 DEBUG : fuse: <- Write [ID=0xb5 Node=0x3 Uid=502 Gid=502 Pid=32351] 0x1 6 @0 fl=0 lock=0 ffl=OpenWriteOnly
2018/12/02 22:06:54 DEBUG : &{test-deadlock/1 (rw)}: Write: len=6, offset=0
2018/12/02 22:06:54 DEBUG : &{test-deadlock/1 (rw)}: >Write: written=6, err=<nil>
2018/12/02 22:06:54 DEBUG : fuse: -> [ID=0xb5] Write 6
2018/12/02 22:06:54 DEBUG : fuse: <- Flush [ID=0xb6 Node=0x3 Uid=502 Gid=502 Pid=32351] 0x1 fl=0x0 lk=0xb8ca17e21032f904
2018/12/02 22:06:54 DEBUG : &{test-deadlock/1 (rw)}: Flush: 
2018/12/02 22:06:54 DEBUG : test-deadlock/1(0xc0000bdb00): close: 
2018/12/02 22:06:54 DEBUG : test-deadlock/1: Couldn't find file - need to transfer
2018/12/02 22:06:54 DEBUG : test-deadlock: invalidating directory cache
2018/12/02 22:06:54 DEBUG : Google drive root 'test': Checking for changes on remote
2018/12/02 22:06:54 DEBUG : Google drive root 'test': Checking for changes on remote
2018/12/02 22:06:55 DEBUG : Google drive root 'test': Checking for changes on remote
2018/12/02 22:06:55 DEBUG : fuse: <- Lookup [ID=0xb7 Node=0x1 Uid=502 Gid=502 Pid=11559] "test-deadlock"
2018/12/02 22:06:55 DEBUG : /: Lookup: name="test-deadlock"
2018/12/02 22:06:55 DEBUG : /: >Lookup: node=test-deadlock/, err=<nil>
2018/12/02 22:06:55 DEBUG : test-deadlock/: Attr: 
2018/12/02 22:06:55 DEBUG : test-deadlock/: >Attr: attr=valid=1s ino=0 size=0 mode=drwxrwxr-x, err=<nil>
2018/12/02 22:06:55 DEBUG : fuse: -> [ID=0xb7] Lookup 0x4 gen=42 valid=1s attr={valid=1s ino=2334695540606880430 size=0 mode=drwxrwxr-x}
2018/12/02 22:06:55 DEBUG : fuse: <- Forget [ID=0xb8 Node=0x4 Uid=0 Gid=0 Pid=0] 1
2018/12/02 22:06:55 DEBUG : fuse: -> [ID=0xb8] Forget
2018/12/02 22:06:55 DEBUG : fuse: <- Forget [ID=0xba Node=0x2 Uid=0 Gid=0 Pid=0] 1
2018/12/02 22:06:55 DEBUG : fuse: <- Lookup [ID=0xb9 Node=0x1 Uid=502 Gid=502 Pid=11559] "test-deadlock"
2018/12/02 22:06:55 DEBUG : fuse: -> [ID=0xba] Forget
2018/12/02 22:06:55 DEBUG : /: Lookup: name="test-deadlock"
2018/12/02 22:06:55 DEBUG : /: >Lookup: node=test-deadlock/, err=<nil>
2018/12/02 22:06:55 DEBUG : test-deadlock/: Attr: 
2018/12/02 22:06:55 DEBUG : test-deadlock/: >Attr: attr=valid=1s ino=0 size=0 mode=drwxrwxr-x, err=<nil>
2018/12/02 22:06:55 DEBUG : fuse: -> [ID=0xb9] Lookup 0x2 gen=43 valid=1s attr={valid=1s ino=2334695540606880430 size=0 mode=drwxrwxr-x}
2018/12/02 22:06:55 DEBUG : fuse: <- Open [ID=0xbb Node=0x2 Uid=502 Gid=502 Pid=11559] dir=true fl=OpenReadOnly+OpenDirectory+OpenNonblock
2018/12/02 22:06:55 DEBUG : fuse: -> [ID=0xbb] Open 0x2 fl=0
2018/12/02 22:06:55 DEBUG : fuse: <- Read [ID=0xbc Node=0x2 Uid=502 Gid=502 Pid=11559] 0x2 4096 @0x0 dir=true fl=0 lock=0 ffl=OpenReadOnly+OpenDirectory+OpenNonblock
2018/12/02 22:06:55 DEBUG : test-deadlock/: ReadDirAll: 
2018/12/02 22:06:55 DEBUG : Google drive root 'test': Checking for changes on remote
2018/12/02 22:06:55 INFO  : test-deadlock/1: Copied (new)
2018/12/02 22:06:55 DEBUG : Google drive root 'test': Checking for changes on remote
2018/12/02 22:07:01 DEBUG : fuse: <- Lookup [ID=0xbd Node=0x1 Uid=502 Gid=502 Pid=11574] "test-deadlock"
2018/12/02 22:07:01 DEBUG : /: Lookup: name="test-deadlock"
2018/12/02 22:07:01 DEBUG : /: >Lookup: node=test-deadlock/, err=<nil>
2018/12/02 22:07:01 DEBUG : test-deadlock/: Attr: 
2018/12/02 22:07:01 DEBUG : test-deadlock/: >Attr: attr=valid=1s ino=0 size=0 mode=drwxrwxr-x, err=<nil>
2018/12/02 22:07:01 DEBUG : fuse: -> [ID=0xbd] Lookup 0x4 gen=44 valid=1s attr={valid=1s ino=2334695540606880430 size=0 mode=drwxrwxr-x}
2018/12/02 22:07:01 DEBUG : fuse: <- Forget [ID=0xbe Node=0x4 Uid=0 Gid=0 Pid=0] 1
2018/12/02 22:07:01 DEBUG : fuse: -> [ID=0xbe] Forget
2018/12/02 22:07:01 DEBUG : fuse: <- Lookup [ID=0xbf Node=0x1 Uid=502 Gid=502 Pid=11574] "test-deadlock"
2018/12/02 22:07:01 DEBUG : /: Lookup: name="test-deadlock"
2018/12/02 22:07:01 DEBUG : /: >Lookup: node=test-deadlock/, err=<nil>
2018/12/02 22:07:01 DEBUG : test-deadlock/: Attr: 
2018/12/02 22:07:01 DEBUG : test-deadlock/: >Attr: attr=valid=1s ino=0 size=0 mode=drwxrwxr-x, err=<nil>
2018/12/02 22:07:01 DEBUG : fuse: -> [ID=0xbf] Lookup 0x4 gen=45 valid=1s attr={valid=1s ino=2334695540606880430 size=0 mode=drwxrwxr-x}
2018/12/02 22:07:01 DEBUG : fuse: <- Open [ID=0xc0 Node=0x4 Uid=502 Gid=502 Pid=11574] dir=true fl=OpenReadOnly+OpenDirectory+OpenNonblock
2018/12/02 22:07:01 DEBUG : fuse: -> [ID=0xc0] Open 0x3 fl=0
2018/12/02 22:07:01 DEBUG : fuse: <- Read [ID=0xc1 Node=0x4 Uid=502 Gid=502 Pid=11574] 0x3 4096 @0x0 dir=true fl=0 lock=0 ffl=OpenReadOnly+OpenDirectory+OpenNonblock
2018/12/02 22:07:01 DEBUG : test-deadlock/: ReadDirAll: 
```

SIGQUIT gives this - 3 goroutines blocked on semacquire
- ForgetPath
- setObjectNoUpdate
- addObject

```
goroutine 14 [semacquire]:
sync.runtime_SemacquireMutex(0xc0006bdd14, 0xc0004f0800)
	/opt/go/go1.11/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc0006bdd10)
	/opt/go/go1.11/src/sync/mutex.go:134 +0xff
github.com/ncw/rclone/vfs.(*Dir).ForgetPath(0xc000358240, 0xc0004f08c0, 0xf, 0x3)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/dir.go:116 +0x173
github.com/ncw/rclone/vfs.(*Dir).ForgetPath-fm(0xc0004f08c0, 0xf, 0x3)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/vfs.go:230 +0x48
github.com/ncw/rclone/backend/drive.(*Fs).changeNotifyRunner(0xc0001be900, 0xc00006c570, 0xc00015e180, 0x7, 0x0, 0x0, 0x0, 0x0)
	/home/ncw/go/src/github.com/ncw/rclone/backend/drive/drive.go:2182 +0xa24
github.com/ncw/rclone/backend/drive.(*Fs).ChangeNotify.func1(0xc0001be900, 0xc0002ce000, 0xc00006c570)
	/home/ncw/go/src/github.com/ncw/rclone/backend/drive/drive.go:2095 +0x2e2
created by github.com/ncw/rclone/backend/drive.(*Fs).ChangeNotify
	/home/ncw/go/src/github.com/ncw/rclone/backend/drive/drive.go:2061 +0x53

goroutine 84 [semacquire]:
sync.runtime_SemacquireMutex(0xc0005ab5fc, 0x14e0900)
	/opt/go/go1.11/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc0005ab5f8)
	/opt/go/go1.11/src/sync/mutex.go:134 +0xff
github.com/ncw/rclone/vfs.(*File).setObjectNoUpdate(0xc0005ab5e0, 0x14e09e0, 0xc0007b4680)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/file.go:343 +0x3a
github.com/ncw/rclone/vfs.(*Dir)._readDirFromEntries(0xc0006bdcb0, 0xc00022e6e0, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/dir.go:250 +0x4e5
github.com/ncw/rclone/vfs.(*Dir)._readDir(0xc0006bdcb0, 0x1387b40, 0xc0006bdd10)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/dir.go:217 +0x124
github.com/ncw/rclone/vfs.(*Dir).ReadDirAll(0xc0006bdcb0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/dir.go:418 +0xa5
github.com/ncw/rclone/cmd/mount.(*Dir).ReadDirAll(0xc00022d5b0, 0x14d4c60, 0xc000174840, 0x0, 0x0, 0x0, 0xe0, 0xa5)
	/home/ncw/go/src/github.com/ncw/rclone/cmd/mount/dir.go:94 +0x193
github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).handleRequest(0xc000774000, 0x14d4c60, 0xc000174840, 0x14c5260, 0xc00022d5b0, 0xc0001e3c80, 0x14d19e0, 0xc0000be000, 0xc0007c5ef8, 0xc0002b7690, ...)
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:1209 +0x3dc
github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).serve(0xc000774000, 0x14d19e0, 0xc0000be000)
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:878 +0x410
github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).Serve.func1(0xc000774000, 0x14d19e0, 0xc0000be000)
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:425 +0x6e
created by github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).Serve
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:423 +0x31e

goroutine 106 [semacquire]:
sync.runtime_SemacquireMutex(0xc0006bdd14, 0xc0007c9800)
	/opt/go/go1.11/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc0006bdd10)
	/opt/go/go1.11/src/sync/mutex.go:134 +0xff
github.com/ncw/rclone/vfs.(*Dir).addObject(0xc0006bdcb0, 0x14e35a0, 0xc0005ab5e0)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/dir.go:187 +0x3a
github.com/ncw/rclone/vfs.(*File).setObject(0xc0005ab5e0, 0x14e09e0, 0xc00016e300)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/file.go:337 +0xac
github.com/ncw/rclone/vfs.(*RWFileHandle).close(0xc00035d3e0, 0x1387b40, 0xc00035d3e8)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/read_write.go:311 +0x740
github.com/ncw/rclone/vfs.(*RWFileHandle).Flush(0xc00035d3e0, 0x0, 0x0)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/read_write.go:350 +0xb8
github.com/ncw/rclone/cmd/mount.(*FileHandle).Flush(0xc00015bde0, 0x14d4c60, 0xc00007fbc0, 0xc0000156d0, 0xc00015bde0, 0x1)
	/home/ncw/go/src/github.com/ncw/rclone/cmd/mount/handle.go:72 +0x116
github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).handleRequest(0xc000774000, 0x14d4c60, 0xc00007fbc0, 0x14c52a0, 0xc00022d888, 0xc00007fa80, 0x14d1660, 0xc0000156d0, 0xc0007c9ef8, 0x553141476b414d4c, ...)
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:1282 +0x2bec
github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).serve(0xc000774000, 0x14d1660, 0xc0000156d0)
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:878 +0x410
github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).Serve.func1(0xc000774000, 0x14d1660, 0xc0000156d0)
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:425 +0x6e
created by github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).Serve
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:423 +0x31e
```

And again
- ForgetPath
- setObjectNoUpdate
- addObject

```
goroutine 332 [semacquire]:
sync.runtime_SemacquireMutex(0xc0003f57dc, 0x14e0900)
	/opt/go/go1.11/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc0003f57d8)
	/opt/go/go1.11/src/sync/mutex.go:134 +0xff
github.com/ncw/rclone/vfs.(*File).setObjectNoUpdate(0xc0003f57c0, 0x14e09e0, 0xc000626780)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/file.go:343 +0x3a
github.com/ncw/rclone/vfs.(*Dir)._readDirFromEntries(0xc0005246c0, 0xc00015e830, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/dir.go:250 +0x4e5
github.com/ncw/rclone/vfs.(*Dir)._readDir(0xc0005246c0, 0x1387b40, 0xc000524720)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/dir.go:217 +0x124
github.com/ncw/rclone/vfs.(*Dir).ReadDirAll(0xc0005246c0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/dir.go:418 +0xa5
github.com/ncw/rclone/cmd/mount.(*Dir).ReadDirAll(0xc00000e068, 0x14d4c60, 0xc00016f840, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/ncw/go/src/github.com/ncw/rclone/cmd/mount/dir.go:94 +0x193
github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).handleRequest(0xc0002f0240, 0x14d4c60, 0xc00016f840, 0x14c5260, 0xc00000e068, 0xc00019a940, 0x14d19e0, 0xc000452000, 0xc000661ef8, 0x0, ...)
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:1209 +0x3dc
github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).serve(0xc0002f0240, 0x14d19e0, 0xc000452000)
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:878 +0x410
github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).Serve.func1(0xc0002f0240, 0x14d19e0, 0xc000452000)
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:425 +0x6e
created by github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).Serve
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:423 +0x31e

goroutine 16 [semacquire]:
sync.runtime_SemacquireMutex(0xc000524724, 0xc0002f4400)
	/opt/go/go1.11/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc000524720)
	/opt/go/go1.11/src/sync/mutex.go:134 +0xff
github.com/ncw/rclone/vfs.(*Dir).ForgetPath(0xc0005a0000, 0xc0002f44c0, 0xf, 0x3)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/dir.go:116 +0x173
github.com/ncw/rclone/vfs.(*Dir).ForgetPath-fm(0xc0002f44c0, 0xf, 0x3)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/vfs.go:230 +0x48
github.com/ncw/rclone/backend/drive.(*Fs).changeNotifyRunner(0xc0000c0d80, 0xc0001afc40, 0xc00043c230, 0x7, 0x0, 0x0, 0x0, 0x0)
	/home/ncw/go/src/github.com/ncw/rclone/backend/drive/drive.go:2182 +0xa24
github.com/ncw/rclone/backend/drive.(*Fs).ChangeNotify.func1(0xc0000c0d80, 0xc0000426c0, 0xc0001afc40)
	/home/ncw/go/src/github.com/ncw/rclone/backend/drive/drive.go:2095 +0x2e2
created by github.com/ncw/rclone/backend/drive.(*Fs).ChangeNotify
	/home/ncw/go/src/github.com/ncw/rclone/backend/drive/drive.go:2061 +0x53

goroutine 293 [semacquire]:
sync.runtime_SemacquireMutex(0xc000524724, 0xc00002d800)
	/opt/go/go1.11/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc000524720)
	/opt/go/go1.11/src/sync/mutex.go:134 +0xff
github.com/ncw/rclone/vfs.(*Dir).addObject(0xc0005246c0, 0x14e35a0, 0xc0003f57c0)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/dir.go:187 +0x3a
github.com/ncw/rclone/vfs.(*File).setObject(0xc0003f57c0, 0x14e09e0, 0xc000626280)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/file.go:337 +0xac
github.com/ncw/rclone/vfs.(*RWFileHandle).close(0xc0000bfe60, 0x1387b40, 0xc0000bfe68)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/read_write.go:311 +0x740
github.com/ncw/rclone/vfs.(*RWFileHandle).Flush(0xc0000bfe60, 0x0, 0x0)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/read_write.go:350 +0xb8
github.com/ncw/rclone/cmd/mount.(*FileHandle).Flush(0xc0001aeda0, 0x14d4c60, 0xc00007f2c0, 0xc0006b46e0, 0xc0001aeda0, 0x1)
	/home/ncw/go/src/github.com/ncw/rclone/cmd/mount/handle.go:72 +0x116
github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).handleRequest(0xc0002f0240, 0x14d4c60, 0xc00007f2c0, 0x14c52a0, 0xc00000e170, 0xc00007f180, 0x14d1660, 0xc0006b46e0, 0xc00002def8, 0xa, ...)
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:1282 +0x2bec
github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).serve(0xc0002f0240, 0x14d1660, 0xc0006b46e0)
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:878 +0x410
github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).Serve.func1(0xc0002f0240, 0x14d1660, 0xc0006b46e0)
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:425 +0x6e
created by github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).Serve
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:423 +0x31e
```

**Comment by trumblejoe** [comment]
Hmm, might the poll-interval be changing the pointer reference of the Lookup?

**Comment by B4dM4n** [comment]
Nice catch.

I tracked down the acquired locks and narrowed it to these two goroutines:
```
goroutine 21465 [semacquire]:
sync.runtime_SemacquireMutex(0xc00055465c, 0x13d9100)
	/nix/store/g0qnlxjw2qs7rk0aamicy8z3x62pcrx8-go-1.11/share/go/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc000554658)
	/nix/store/g0qnlxjw2qs7rk0aamicy8z3x62pcrx8-go-1.11/share/go/src/sync/mutex.go:134 +0xff
github.com/ncw/rclone/vfs.(*File).setObjectNoUpdate(0xc000554640, 0x13d91c0, 0xc000592000)
	/home/fabian/Sync/Projects/Go/src/github.com/ncw/rclone/vfs/file.go:343 +0x3a
  **f.mu.Lock()**
github.com/ncw/rclone/vfs.(*Dir)._readDirFromEntries(0xc000314240, 0xc0000ea040, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/fabian/Sync/Projects/Go/src/github.com/ncw/rclone/vfs/dir.go:250 +0x4e5
github.com/ncw/rclone/vfs.(*Dir)._readDir(0xc000314240, 0x1290e98, 0xc0003142a0)
	/home/fabian/Sync/Projects/Go/src/github.com/ncw/rclone/vfs/dir.go:217 +0x124
github.com/ncw/rclone/vfs.(*Dir).ReadDirAll(0xc000314240, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/fabian/Sync/Projects/Go/src/github.com/ncw/rclone/vfs/dir.go:418 +0xa5
  **d.mu.Lock()**
github.com/ncw/rclone/cmd/mount.(*Dir).ReadDirAll(0xc0004d40e8, 0x13cd9e0, 0xc0005dc1c0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/fabian/Sync/Projects/Go/src/github.com/ncw/rclone/cmd/mount/dir.go:94 +0x193

goroutine 21430 [semacquire]:
sync.runtime_SemacquireMutex(0xc0003142a4, 0xc00103b800)
	/nix/store/g0qnlxjw2qs7rk0aamicy8z3x62pcrx8-go-1.11/share/go/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc0003142a0)
	/nix/store/g0qnlxjw2qs7rk0aamicy8z3x62pcrx8-go-1.11/share/go/src/sync/mutex.go:134 +0xff
github.com/ncw/rclone/vfs.(*Dir).addObject(0xc000314240, 0x13dbc20, 0xc000554640)
	/home/fabian/Sync/Projects/Go/src/github.com/ncw/rclone/vfs/dir.go:187 +0x3a
  **d.mu.Lock()**
github.com/ncw/rclone/vfs.(*File).setObject(0xc000554640, 0x13d91c0, 0xc0018fa400)
	/home/fabian/Sync/Projects/Go/src/github.com/ncw/rclone/vfs/file.go:337 +0xac
  **f.mu.Lock()**
github.com/ncw/rclone/vfs.(*RWFileHandle).close(0xc0007b57a0, 0x1290e98, 0xc0007b57a8)
	/home/fabian/Sync/Projects/Go/src/github.com/ncw/rclone/vfs/read_write.go:311 +0x740
  fh.file.muRW.Lock()
github.com/ncw/rclone/vfs.(*RWFileHandle).Flush(0xc0007b57a0, 0x0, 0x0)
	/home/fabian/Sync/Projects/Go/src/github.com/ncw/rclone/vfs/read_write.go:350 +0xb8
  fh.mu.Lock()
github.com/ncw/rclone/cmd/mount.(*FileHandle).Flush(0xc00057fb80, 0x13cd9e0, 0xc0004f5500, 0xc000536f00, 0xc00057fb80, 0x1)
	/home/fabian/Sync/Projects/Go/src/github.com/ncw/rclone/cmd/mount/handle.go:72 +0x116
```
I added lines containing the held locks to each call and am pretty sure the ** marked ones are causing the deadlock.

During `(*RWFileHandle).Flush` the file tries to add itself to it's directory using `(*Dir).addObject` which tries to acquire `d.mu.Lock()` while holding `f.mu.Lock()`.
Concurrently `(*Dir).ReadDirAll` tries to set the file object with `(*File).setObjectNoUpdate` which acquires `f.mu.Lock()` while holding `d.mu.Lock()`.

So `poll-interval` only increases the chance of this happening by invalidating the cache with `(*Dir).ForgetPath`.

The best solution is probably to disable `(*File).setObjectNoUpdate` while the `RWFileHandle` is being closed, as it would be possible to override the uploaded object with a old version during `(*Dir).ReadDirAll` otherwise.

**Comment by ncw** [comment]
I managed to provoke two kinds of deadlocks in the end. Here is the second

```
goroutine 10 [semacquire, 216 minutes]:
sync.runtime_SemacquireMutex(0xc000614334, 0xc0004b6f00)
	/opt/go/go1.11/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc000614330)
	/opt/go/go1.11/src/sync/mutex.go:134 +0xff
github.com/ncw/rclone/vfs.(*Dir).ForgetPath(0xc0006f0360, 0xc0004b6f60, 0xf, 0x3)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/dir.go:116 +0x173
github.com/ncw/rclone/vfs.(*Dir).ForgetPath-fm(0xc0004b6f60, 0xf, 0x3)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/vfs.go:230 +0x48
github.com/ncw/rclone/backend/drive.(*Fs).changeNotifyRunner(0xc00008f200, 0xc00023e910, 0xc0000b8e80, 0x7, 0x0, 0x0, 0x0, 0x0)
	/home/ncw/go/src/github.com/ncw/rclone/backend/drive/drive.go:2182 +0xa24
github.com/ncw/rclone/backend/drive.(*Fs).ChangeNotify.func1(0xc00008f200, 0xc000042900, 0xc00023e910)
	/home/ncw/go/src/github.com/ncw/rclone/backend/drive/drive.go:2095 +0x2e2
created by github.com/ncw/rclone/backend/drive.(*Fs).ChangeNotify
	/home/ncw/go/src/github.com/ncw/rclone/backend/drive/drive.go:2061 +0x53

goroutine 18812 [semacquire, 216 minutes]:
sync.runtime_SemacquireMutex(0xc000614334, 0xc000542500)
	/opt/go/go1.11/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc000614330)
	/opt/go/go1.11/src/sync/mutex.go:134 +0xff
github.com/ncw/rclone/vfs.(*Dir).delObject(0xc0006142d0, 0x1ea5c12, 0x1)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/dir.go:194 +0x36
github.com/ncw/rclone/vfs.(*File).Remove(0xc00092f9a0, 0x0, 0x0)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/file.go:459 +0xf7
github.com/ncw/rclone/vfs.(*Dir).RemoveName(0xc0006142d0, 0x1ea5c12, 0x1, 0x6, 0xc0002229f0)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/dir.go:540 +0x119
github.com/ncw/rclone/cmd/mount.(*Dir).Remove(0xc0000bcba8, 0x14d61c0, 0xc00007fd80, 0xc00018c230, 0xc0000bcba8, 0x1)
	/home/ncw/go/src/github.com/ncw/rclone/cmd/mount/dir.go:148 +0x185
github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).handleRequest(0xc000606000, 0x14d61c0, 0xc00007fd80, 0x14c67c0, 0xc0000bcba8, 0xc0021d7880, 0x14d3000, 0xc00018c230, 0xc0007a3ef8, 0xa, ...)
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:1024 +0x3499
github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).serve(0xc000606000, 0x14d3000, 0xc00018c230)
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:878 +0x410
github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).Serve.func1(0xc000606000, 0x14d3000, 0xc00018c230)
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:425 +0x6e
created by github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).Serve
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:423 +0x31e

goroutine 18858 [semacquire, 216 minutes]:
sync.runtime_SemacquireMutex(0xc00092f9bc, 0x14e1f00)
	/opt/go/go1.11/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc00092f9b8)
	/opt/go/go1.11/src/sync/mutex.go:134 +0xff
github.com/ncw/rclone/vfs.(*File).setObjectNoUpdate(0xc00092f9a0, 0x14e1f40, 0xc00013bc00)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/file.go:344 +0x3a
github.com/ncw/rclone/vfs.(*Dir)._readDirFromEntries(0xc0006142d0, 0xc00001c2e0, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/dir.go:250 +0x4e5
github.com/ncw/rclone/vfs.(*Dir)._readDir(0xc0006142d0, 0x1389018, 0xc000614330)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/dir.go:217 +0x124
github.com/ncw/rclone/vfs.(*Dir).stat(0xc0006142d0, 0x1ea5c11, 0x1, 0x0, 0x0, 0x0, 0x0)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/dir.go:323 +0x7f
github.com/ncw/rclone/vfs.(*Dir).Stat(0xc0006142d0, 0x1ea5c11, 0x1, 0x10, 0xc000326140, 0x2, 0x2)
	/home/ncw/go/src/github.com/ncw/rclone/vfs/dir.go:402 +0x43
github.com/ncw/rclone/cmd/mount.(*Dir).Lookup(0xc000726700, 0x14d61c0, 0xc000810180, 0xc0000c1b80, 0xc0004d8240, 0xc0004d6301, 0x5, 0x5, 0x0)
	/home/ncw/go/src/github.com/ncw/rclone/cmd/mount/dir.go:73 +0x1d0
github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).handleRequest(0xc000606000, 0x14d61c0, 0xc000810180, 0x14c67c0, 0xc000726700, 0xc000810140, 0x14d2e40, 0xc0000c1b80, 0xc0007a9ef8, 0xa, ...)
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:1050 +0x31cb
github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).serve(0xc000606000, 0x14d2e40, 0xc0000c1b80)
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:878 +0x410
github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).Serve.func1(0xc000606000, 0x14d2e40, 0xc0000c1b80)
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:425 +0x6e
created by github.com/ncw/rclone/vendor/bazil.org/fuse/fs.(*Server).Serve
	/home/ncw/go/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fs/serve.go:423 +0x31e
```

I've worked around them by narrowing the locking so the file lock and the directory lock don't overlap in File.setObject and File.Remove.  I think that is OK but I'm not 100% sure!

I'd really appreciate a review of this @B4dM4n if you have time :-) 5cc92ab0f383e9780cc995caa331c5f287c11cfd I think your analysis is spot on - it agrees 100% with mine.  I chose a different solution though - what do you think of that?  Note that `f.d` is read only and hence doesn't need to be protected with the mutex.

@darthShadow here is a beta for you to try

https://beta.rclone.org/branch/v1.45-021-g5cc92ab0-fix-2811-deadlock-beta/ (uploaded in 15-30 mins)


### deadlock 1

```
ReadWrite.Close()
-> file.Flush
-> file.setObject()
   -> f.mu.Lock
     -> d.addObject
         -> d.mu.Lock (DEADLOCK)

Dir.ReadDirAll()
-> d.mu.Lock
  -> _readDir
  -> _readDirFromEntries
  -> file.setObjectNoUpdate
      -> f.mu.Lock (DEADLOCK)

ChangeNotify
-> ForgetPath
  -> dir.mu.Lock (DEADLOCK)
```

### deadlock 2

```
Dir.Remove
-> file.Remove
   -> f.mu.Lock
   -> file.delObject()
     -> d.mu.Lock (DEADLOCK)

Dir.Lookup
Dir.Stat
dir.stat
-> d.mu.Lock
  -> _readDir
  -> _readDirFromEntries
  -> file.setObjectNoUpdate
      -> f.mu.Lock (DEADLOCK)

ChangeNotify
-> ForgetPath
  -> dir.mu.Lock (DEADLOCK)
```

**Comment by darthShadow** [comment]
@ncw Did something fail in the build? I don't see builds for any system other than Windows.

**Comment by ncw** [comment]
@darthShadow I've fixed the build (hopefully!) here:

https://beta.rclone.org/branch/v1.45-022-g86ed4e32-fix-2811-deadlock-beta/ (uploaded in 15-30 mins)

**Comment by Nottt** [comment]
I believe I'm suffering from deadlocks when using the rclone cache against a glusterfs remote with very low latency.

How can I debug it? The logs at INFO don't log anything, and this happen randomly...a debug log would be >10gb for sure

**Comment by ncw** [comment]
@Nott If it is a deadlock then `kill -QUIT` the rclone process and collect the backtrace, or use [this procedure](https://rclone.org/rc/#debugging-go-routine-leaks) via the rc.

If you want more help then please make a new issue with that in :-)

Note glusterfs is weird... Its nanosecond timestamps aren't consistent which could conceivably mess with rclone so using `--modify-window` might be a good idea.

**Comment by Nottt** [comment]
The rclone mount is read-only, I add stuff to glusterfs then refresh rclone with vfs/refresh.  Do you still think I should use modify window?

I'll try killing rclone next time this happens...but rclone keeps working when this happens...what actually happens is that the php scripts accessing the mount get stuck in "D" state and eventually the kernel logs it with hung_task_timeout.

**Comment by ncw** [comment]
> The rclone mount is read-only, I add stuff to glusterfs then refresh rclone with vfs/refresh. Do you still think I should use modify window?

I don't know... Hold off on that until we have a backtrace

> I'll try killing rclone next time this happens...but rclone keeps working when this happens...what actually happens is that the php scripts accessing the mount get stuck in "D" state and eventually the kernel logs it with hung_task_timeout.

That sounds very much like a deadlock. I'll need to know exactly which rclone version you are running in the bugreport too!

**Comment by Nottt** [comment]
None of my servers had this yet. I have found that 2 sysctl settings managed to delay the deadlocks by a lot:

```
vm.dirty_background_ratio = 5 
vm.dirty_ratio = 80
```

After I used this, my servers stay much longer without having the issue... before it would be very frequent, now it happens like once a week or so

**Comment by ncw** [comment]
Great - thanks. WIth a backtrace these things are easy to fix!
